atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/GetRowsColumnRangeIteratorTest.java

```diff
@@ -0,0 +1,157 @@
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.atlasdb.keyvalue.impl.InMemoryKeyValueService;
+
+@RunWith(MockitoJUnitRunner.class)
+public class GetRowsColumnRangeIteratorTest {
+
+    private static final TableReference TABLE_REFERENCE = TableReference.createWithEmptyNamespace("test");
+    private static final byte[] ROW = "row".getBytes(StandardCharsets.UTF_8);
+    private static final int BATCH_SIZE = 10;
+    public static final BatchColumnRangeSelection COLUMN_RANGE_SELECTION =
+            BatchColumnRangeSelection.create(null, null, BATCH_SIZE);
+
+    private final KeyValueService kvs = new InMemoryKeyValueService(true);
+    private final ColumnRangeBatchProvider batchProvider = new ColumnRangeBatchProvider(
+            kvs,
+            TABLE_REFERENCE,
+            ROW,
+            COLUMN_RANGE_SELECTION,
+            Long.MAX_VALUE);
+
+    @Test
+    public void ifBatchIsEmptyNoValidateCallsAreMade() {
+        Runnable validationStep = mock(Runnable.class);
+        Iterator<Map.Entry<Cell, byte[]>> iterator = createIteratorUnderTest(validationStep);
+
+        List<Map.Entry<Cell, byte[]>> entries = ImmutableList.copyOf(iterator);
+
+        assertThat(entries).isEmpty();
+        verifyNoInteractions(validationStep);
+    }
+
+    @Test
+    public void firstBatchHasNoValidation() {
+        Runnable validationStep = mock(Runnable.class);
+        Set<Cell> puts = putColumns(BATCH_SIZE + 5);
+
+        int limit = BATCH_SIZE - 1;
+        Iterator<Map.Entry<Cell, byte[]>> iteratorUnderTest = createIteratorUnderTest(validationStep);
+
+        // still under the first batch size limit
+        List<Map.Entry<Cell, byte[]>> firstBatchBarOne =
+                ImmutableList.copyOf(Iterators.limit(iteratorUnderTest, limit));
+        verifyNoInteractions(validationStep);
+
+        // the last element in the first batch is still on the first batch
+        Map.Entry<Cell, byte[]> lastInFirstBatch = Iterators.getNext(iteratorUnderTest, null);
+        assertThat(lastInFirstBatch).isNotNull();
+        verifyNoInteractions(validationStep);
+
+        // consume one more i.e. batch size amount
+        Map.Entry<Cell, byte[]> firstInSecondBatch = Iterators.getNext(iteratorUnderTest, null);
+        assertThat(firstInSecondBatch).isNotNull();
+        verify(validationStep, times(1)).run();
+
+        // validation step is called per batch
+        ImmutableList<Map.Entry<Cell, byte[]>> restOfSecondBatch = ImmutableList.copyOf(iteratorUnderTest);
+        verifyNoMoreInteractions(validationStep);
+
+        Map<Cell, byte[]> returnedEntries = ImmutableMap.<Cell, byte[]>builder()
+                .putAll(firstBatchBarOne)
+                .put(lastInFirstBatch)
+                .put(firstInSecondBatch)
+                .putAll(restOfSecondBatch)
+                .build();
+
+
+        assertThat(returnedEntries.keySet())
+                .as("we can read all that we wrote")
+                .isEqualTo(puts);
```

Comment:
Might be good to check the ordering here (we can make a `SortedSet` copy of this - `Cell` compare is based on the byte order of rows and columns - and then build a List rather than a Map of returnedEntries and check the ordering)

Reply:
maybe something like
```
        List<Cell> cellsReadInOrder = Lists.newArrayList();
        cellsReadInOrder.addAll(firstBatchBarOne.stream().map(Map.Entry::getKey).collect(Collectors.toList()));
        cellsReadInOrder.add(lastInFirstBatch.getKey());
        cellsReadInOrder.add(firstInSecondBatch.getKey());
        cellsReadInOrder.addAll(restOfSecondBatch.stream().map(Map.Entry::getKey).collect(Collectors.toList()));

        assertThat(cellsReadInOrder)
                .as("we can read all that we wrote")
                .hasSameElementsAs(puts);
        assertThat(cellsReadInOrder)
                .as("iterator returns cells back in sorted order")
                .isSorted();
```